objects/src/builder/mod.rs

```diff
@@ -0,0 +1,274 @@
+    }
+
+    pub fn build_fungible(&self) -> Result<FungibleAsset, AssetError> {
+        FungibleAsset::new(self.faucet_id, self.amount)
+    }
+}
+
+/// Builder for an `NonFungibleAssetDetails`, the builder can be configured and used multipled times.
+pub struct NonFungibleAssetDetailsBuilder<T: Rng> {
+    faucet_id: AccountId,
+    rng: T,
+}
+
+impl<T: Rng> NonFungibleAssetDetailsBuilder<T> {
+    pub fn new(faucet_id: AccountId, rng: T) -> Result<Self, AssetError> {
+        if !matches!(faucet_id.account_type(), AccountType::NonFungibleFaucet) {
+            return Err(AssetError::not_a_non_fungible_faucet_id(faucet_id));
+        }
+
+        Ok(Self { faucet_id, rng })
+    }
+
+    pub fn build(&mut self) -> Result<NonFungibleAssetDetails, AssetError> {
+        let data = (&mut self.rng).sample_iter(Standard).take(5).collect();
+        NonFungibleAssetDetails::new(self.faucet_id, data)
+    }
+}
+
+/// Builder for an `NonFungibleAsset`, the builder can be configured and used multipled times.
+pub struct NonFungibleAssetBuilder<T: Rng> {
+    details_builder: NonFungibleAssetDetailsBuilder<T>,
+}
+
+impl<T: Rng> NonFungibleAssetBuilder<T> {
+    pub fn new(faucet_id: AccountId, rng: T) -> Result<Self, AssetError> {
+        let details_builder = NonFungibleAssetDetailsBuilder::new(faucet_id, rng)?;
+        Ok(Self { details_builder })
+    }
+
+    pub fn with_details(
+        &mut self,
+        details: &NonFungibleAssetDetails,
+    ) -> Result<NonFungibleAsset, AssetError> {
+        NonFungibleAsset::new(&details)
+    }
+
+    pub fn build(&mut self) -> Result<NonFungibleAsset, AssetError> {
+        let details = self.details_builder.build()?;
+        NonFungibleAsset::new(&details)
+    }
+}
+
+/// Builder for an `Account`, the builder allows for a fluent API to construct an account. Each
+/// account needs a unique builder.
+pub struct AccountBuilder<T: Rng> {
+    assets: Vec<Asset>,
+    storage_builder: AccountStorageBuider,
+    code: String,
+    nonce: Felt,
+    account_id_builder: AccountIdBuilder<T>,
+}
+
+const DEFAULT_ACCOUNT_CODE: &str = "\
+use.miden::sat::account
+
+export.incr_nonce
+    push.0 swap
+    # => [value, 0]
+
+    exec.account::incr_nonce
+    # => [0]
+end
+
+export.set_item
+    exec.account::set_item
+    # => [R', V, 0, 0, 0]
+
+    movup.8 drop movup.8 drop movup.8 drop
+    # => [R', V]
+end
+
+export.set_code
+    padw swapw
+    # => [CODE_ROOT, 0, 0, 0, 0]
+
+    exec.account::set_code
+    # => [0, 0, 0, 0]
+end
+
+export.account_procedure_1
+    push.1.2
+    add
+end
+
+export.account_procedure_2
+    push.2.1
+    sub
+end
+";
+
+impl<T: Rng> AccountBuilder<T> {
+    pub fn new(rng: T) -> Self {
+        Self {
+            assets: vec![],
+            storage_builder: AccountStorageBuider::new(),
+            code: DEFAULT_ACCOUNT_CODE.to_string(),
+            nonce: ZERO,
+            account_id_builder: AccountIdBuilder::new(rng),
+        }
+    }
+
```

Comment:
In some testing cases we may have some static `AccountId`'s that we want to create accounts for.  I wonder if we should also allow `Account`'s to be built against these Id's via setter (`pub fn set_id(..)`) instead of always delegating to the `AccountIdBuilder` to create one for us.

Reply:
That was the initial design, this was removed because of https://github.com/0xPolygonMiden/miden-base/pull/206#discussion_r1303687545

We either get to inject an account id, or we get consistent account ids

Reply:
I wonder if we can modify the implementation to look something like this to support both:

```rust
impl<T: Rng> AccountBuilder<T> {
    pub fn new(rng: T) -> Self {
        Self {
            id: None,
            seed: None,
            assets: vec![],
            storage_builder: AccountStorageBuider::new(),
            code: DEFAULT_ACCOUNT_CODE.to_string(),
            nonce: ZERO,
            account_id_builder: AccountIdBuilder::new(rng),
        }
    }

    pub fn set_id(..) -> Self {
        ...
    }

    pub fn set_seed(..) -> Self {
        ...
    }
    
    ...
    
    pub fn build(mut self) -> Result<Account, AccountError> {
        let account_id = match (&self.id, &self.seed, &self.nonce) {
            (Some(id), None, 0) => panic!("seed must be provided for new account"),
            (Some(id), Some(seed), 0) => {
                // validate seed here
                id
            }
            (Some(id), _, _) => id,
            (None, _, _) => self.account_id_builder.build()?,
        };
        let vault = AccountVault::new(&self.assets)?;
        let storage = self.storage_builder.build();
        let assembler = assembler();
        let account_module_ast = ModuleAst::parse(&self.code).unwrap();
        let account_code = AccountCode::new(account_id, account_module_ast, &assembler)?;
        Ok(Account::new(account_id, vault, storage, account_code, self.nonce))
    }
}
```

Reply:
I mean, we can, but that way you can use `set_id` to set the id that doesn't correspond to the given code with nonce `0`. I'm fine with either way, we just need to decide if we think that we want go guarantee that the id is consistent or not.

Reply:
The idea would be that you validate the seed as part of the build method.  See the code comment I left in the code block above `// validate seed here`, this would basically be equivalent to calling `AccountId::new(seed, code_root, storage_root)` and asserting the returned `AccountId` matches the one provided.  Once again not a priority but something to consider.